File: apps/api-rs/src/models/rag.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::trace::RetrievalSource;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetrievedChunk {
    pub chunk_id: Uuid,
    pub doc_id: Uuid,
    pub doc_title: String,
    pub file_type: String,
    pub content: String,
    pub heading_path: Vec<String>,
    pub page_range: Vec<i32>,
    pub block_ids: Vec<Uuid>,
    pub table_ids: Vec<Uuid>,
    pub anchor_ids: Vec<Uuid>,
    pub primary_anchor_id: Option<Uuid>,
    pub anchor_quality: String,
    pub primary_anchor: Option<crate::models::SourceAnchor>,
    pub metadata: serde_json::Value,
    pub score: f64,
    pub source: RetrievalSource,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RerankedChunk {
    pub chunk: RetrievedChunk,
    pub score: f64,
    pub rank: i32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EvidencePack {
    pub chunks: Vec<RerankedChunk>,
    pub context_text: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetrievalInput {
    pub tenant_id: Uuid,
    pub effective_kb_ids: Vec<Uuid>,
    pub queries: Vec<String>,
    #[serde(default)]
    pub hypothetical_answer: Option<String>,
    pub top_k: usize,
    pub dense_top_k: usize,
    pub bm25_top_k: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetrievalOutput {
    pub chunks: Vec<RetrievedChunk>,
    #[serde(default)]
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RerankInput {
    pub query: String,
    pub chunks: Vec<RetrievedChunk>,
    pub top_k: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContextInput {
    pub chunks: Vec<RerankedChunk>,
    pub original_query: String,
    pub max_context_chars: usize,
}

File: apps/api-bun/src/models/rag.ts
// 移植自 apps/api-rs/src/models/rag.rs
import type { RetrievalSource } from './trace.ts';
import type { SourceAnchor } from './source_anchor.ts';

export interface RetrievedChunk {
  chunk_id: string; doc_id: string; doc_title: string; file_type: string; content: string;
  heading_path: string[]; page_range: number[]; block_ids: string[]; table_ids: string[];
  anchor_ids: string[]; primary_anchor_id: string | null; anchor_quality: string;
  primary_anchor: SourceAnchor | null; anchors: SourceAnchor[]; metadata: Record<string, unknown>;
  score: number; source: RetrievalSource;
}
export interface RerankedChunk { chunk: RetrievedChunk; score: number; rank: number; }
export interface EvidencePack { chunks: RerankedChunk[]; context_text: string; }

export interface RetrievalInput {
  tenant_id: string; effective_kb_ids: string[]; queries: string[];
  hypothetical_answer?: string | null; top_k: number; dense_top_k: number; bm25_top_k: number;
}
export interface RetrievalOutput { chunks: RetrievedChunk[]; warnings?: string[]; }
export interface RerankInput { query: string; chunks: RetrievedChunk[]; top_k: number; }
export interface ContextInput { chunks: RerankedChunk[]; original_query: string; max_context_chars: number; }
